src/7-to-8/major-changes/compatibility-mode.rst

```diff
@@ -39,22 +40,101 @@ The old ``suite.rc`` filename triggers a backward compatibility mode in which:
     :ref:`*optional* <User Guide Optional Outputs>` by the new ``?`` syntax)
 
 
+.. _compat_required_changes:
+
 Required Changes
 ----------------
 
 Providing your Cylc 7 workflow does not use syntax that was deprecated at Cylc 7,
-you may be able to run it with Cylc 8 without any modifications to begin with.
+you may be able to run it using Cylc 8 without any modifications while in
+compatibility mode.
+
+First, run ``cylc validate`` **with Cylc 7** on your ``suite.rc`` workflow
+to check for deprecation warnings and fix those before validating with Cylc 8.
+See :ref:`below <compat.eg.c7val>` for an example.
+
+.. warning::
+
+   ``cylc validate`` operates on the processed ``suite.rc``, which
+   means it will not detect any deprecated syntax that is inside a
+   currently-unused Jinja2/EmPy ``if...else`` branch.
+
+Some workflows may require modifications to either upgrade to Cylc 8 or make
+interoperable with Cylc 8 backward compatibility mode. Read on for more details.
+
+
+Cylc commands in task scripts
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Check for any use of Cylc commands in task scripting. Some Cylc 7 commands
+have been removed and some others now behave differently.
+However, ``cylc message`` and ``cylc broadcast`` have *not* changed.
+See the :ref:`full list of command line interface changes<MajorChangesCLI>`
+and see :ref:`below <compat.eg.cylc-commands>` for an example.
+
+
+Python 2 to 3
+^^^^^^^^^^^^^
+
+Whereas Cylc 7 runs using Python 2, Cylc 8 runs using Python 3. This affects:
+- modules imported in Jinja2
+- Jinja2 filters, tests and globals
+- custom xtrigger functions
+
+Note that task scripts are not affected - they run in an independent
+environment.
+
+See :ref:`py23` for more information and examples of how to implement
+interoperability if your workflows extend Cylc or Jinja2 with custom Python scripts.
+
+
+Restarting a Cylc 7 workflow
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Cylc 8 cannot *restart* a Cylc 7 workflow mid-run. Instead, :ref:`install
+<Installing-workflows>` the workflow to a new run directory and start it
+from scratch at the right cycle point or task(s):
 
-To see if your workflow is compatible with Cylc 8, run ``cylc validate``
-**using Cylc 7**.
+- ``cylc play --start-cycle-point=<cycle>`` (c.f. Cylc 7 *warm start*), or
+- ``cylc play --start-task=<cycle/task>``   (Cylc 8 can start anywhere in the graph)
 
-If tasks in your workflow call Cylc commands directly it might be necessary to
-modify them to be compatible with command line interface changes
-(unfortunately, this is not possible to detect at validation).
+.. note::
+
+   Any previous-cycle workflow data needed by the new run will need to be
+   manually copied over from the original run directory.
+
+
```

Comment:
It will be a very common requirement though. Should we not provide advice on how to do it? 

Maybe don't call it "restarting", to avoid confusion.  Perhaps something like **How to start your migrated workflow from where Cylc 7 left it**. And give general advice and caveats.

Reply:
Or `"Restarting" a Cylc 7 workflow`?

Reply:
> It will be a very common requirement though.

I'm not aware of any requirement or desire at our end. I think the only use case would be really, really, long-running workflows, I'm not aware that we have any examples of this. If they exist I would advise finishing the workflow run with Cylc 7 rather than to attempting to migrate to Cylc 8 part-way through a run.

Migrating like this is hazardous, e.g. remote installation works differently, data might need to be relocated on multiple filesystems, etc. We discussed this in a meeting at our end?

Reply:
I've replaced this subsection with a bullet point saying you can't restart, use Cylc 7 to finish a Cylc 7 workflow

Reply:
> I'm not aware of any requirement or desire at our end. I think the only use case would be really, really, long-running workflows, I'm not aware that we have any examples of this.

Most of our workflows are "really, really, long-running" - i.e. they no final cycle point, or a very distant one. And most are "warm cycling" - i.e there are inter-cycle dependencies, and doing a cold start from scratch is not something you want to do unless you have no choice.

Anyhow, we may have got our wires crossed here. I just meant "advise users how to continue a warm-cycling Cylc 7 workflow with Cylc 8", not a literal in-place restart from the DB. 

Along the lines of:

*Shut down (with Cylc 7) after a particular cycle point, copy any inter-cycle data over to a new Cylc 8 install (taking account of the run-name if appropriate), and cold start that with Cylc 8 at the next cycle point.*